signer-service/src/api/services/siwe.service.js

```diff
@@ -1,75 +1,139 @@
+
+  if (!siweMessage) {
+    throw new Error('Message must be provided as a parameter if it has not been initially validated.');
   }
 
+  // Verify the integrity of the message
   const publicClient = createPublicClient({
     chain: polygon,
     transport: http(),
   });
 
   const valid = await publicClient.verifyMessage({
-    address: maybeSiweData.address,
-    message: maybeSiweData.siweMessage.toMessage(),
+    address: address,
+    message: siweMessage.toMessage(), // Validation must be done on the message as string
     signature,
   });
 
   if (!valid) {
-    throw new Error('Invalid signature.');
+    throw new ValidationError('Invalid signature');
+  }
+
+  // Perform additional checks to ensure message fields
+  if (siweMessage.nonce !== nonce) {
+    throw new ValidationError('Nonce mismatch');
+  }
+
+  if (siweMessage.expirationTime && new Date(siweMessage.expirationTime) < new Date()) {
+    throw new ValidationError('Message has expired');
+  }
+
+  return siweMessage;
+};
+
+// Since the message is created in the UI, we need to verify the fields of the message
+const verifyMessageFields = (siweMessage) => {
+  // Fields we validate on initial
+  const domain = siweMessage.domain;
+  const uri = siweMessage.uri;
+  const scheme = siweMessage.scheme; // must be https
+  const chainId = siweMessage.chainId;
+  const expirationTime = siweMessage.expirationTime;
+
+  if (!VALID_SIWE_LOGIN_ORIGINS.includes(uri)) {
+    throw new ValidationError('Origin not in the list of allowed origins');
   }
 
-  // Perform additional checks to ensure message integrity
-  if (maybeSiweData.siweMessage.nonce !== nonce) {
-    throw new Error('Nonce mismatch.');
+  if (!VALID_SIWE_DOMAINS.includes(domain)) {
+    throw new ValidationError('Incorrect domain');
   }
 
-  if (maybeSiweData.expirationTime && new Date(maybeSiweData.expirationTime) < new Date()) {
-    throw new Error('Message has expired.');
+  if (!VALID_SIWE_CHAINS.includes(chainId)) {
+    throw new ValidationError('Incorrect chain ID');
   }
 
-  // Successful verification, remove messages with same address
-  // from map except for the one we just verified.
+  if (scheme !== 'https') {
+    throw new ValidationError('Scheme must be https');
+  }
+
+  if (expirationTime === NaN || expirationTime === undefined) {
+    throw new ValidationError('Must defined valid expiration time');
+  }
+
+  // Check if expiration is within a reasonable range from current time
+  const currentTime = new Date().getTime();
+  const expirationTimestamp = new Date(expirationTime).getTime();
+  console.log('expirationTimestamp', expirationTimestamp);
+
+  const expirationGracePeriod = 1000 * 60; // 1 minute
+  const expirationPeriodMs = DEFAULT_LOGIN_EXPIRATION_TIME_HOURS * 60 * 60 * 1000;
+  const expectedMinExpirationTimestamp = currentTime + expirationPeriodMs - expirationGracePeriod;
+  const expectedMaxExpirationTimestamp = currentTime + expirationPeriodMs;
```

Comment:
So if I understand this correctly, you want to allow a range of 7days-1min < x < 7days for the expiration period? 

Wouldn't this check fail oftentimes for siwe messages that are not fresh? Let's say I am a user and I have an existing Siwe message that is valid for 7 days. If I authenticate with it one day after creating it, this logic here will throw an error because it's not within the min and max.

Reply:
I think we can't really check for the min and max expiration time if we don't know the timestamp of the creation. 
However, generally this is more a problem of the user interface. If the user interface decides to create very short-lived siwe sessions, so be it. It's a bad UX but that's up to the developer of the frontend.

Reply:
>  7days-1min < x < 7days

Yes exactly. Keep in mind that this check is done only on validation (initial verify and store) of the cookie. After that, the `verifyMessageFields` is not used, and we just check the expiration itself of the cookie and also of the signature/token itself.

The 1 min would then be the max delay that the UI has for sending the crafted message. This is all such that the message would be similar to one created on the backend.

But for the last point, you're right. The most important check is x < 7 days that will check a too-long expiration time. It's not a security concern that a UI sends short messages, but I think is good for being consistent.

Reply:
Oh right, I missed that it's only called by `verifyAndStoreSiweMessage()`. I'm okay if we keep it as is then but let's maybe rename `verifyMessageFields()` to `verifyInitialMessageFields()` to make the difference to `verifySiweMessage()` clearer 😬